core/types/calls.cc

```diff
@@ -1227,9 +1235,43 @@ TypePtr AppliedType::getCallArguments(const GlobalState &gs, NameRef name) const
     return getMethodArguments(gs, klass, name, targs);
 }
 
+bool MetaType::canCallNew(const GlobalState &gs) const {
+    if (isa_type<OrType>(wrapped) || isa_type<AndType>(wrapped)) {
+        return false;
+    }
+
+    if (isa_type<ClassType>(wrapped)) {
+        auto sym = cast_type_nonnull<ClassType>(wrapped).symbol;
+        if (sym == Symbols::untyped() || sym == Symbols::bottom()) {
+            return false;
+        }
+    }
+
+    if (auto *appliedType = cast_type<AppliedType>(wrapped)) {
+        if (appliedType->klass.data(gs)->isSingletonClass(gs)) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
 DispatchResult MetaType::dispatchCall(const GlobalState &gs, const DispatchArgs &args) const {
     switch (args.name.rawId()) {
         case Names::new_().rawId(): {
+            if (!canCallNew(gs)) {
+                auto callLoc = core::Loc(args.locs.file, args.locs.call);
+                if (auto e = gs.beginError(callLoc, errors::Infer::MetaTypeDispatchCall)) {
+                    e.setHeader("Cannot call `{}` on type `{}`", Names::new_().show(gs), wrapped.show(gs));
```

Comment:
ooc, why did you chose a different error message from below?

```suggestion
                    e.setHeader("Call to method `{}` on `{}` mistakes a type for a value", Names::new_().show(gs),
                                this->wrapped.show(gs));
```

Reply:
No particular reason! I think I had it in my head that "`new` isn't a method" but that's, well, wrong. Will change as suggested.